### PR TITLE
Imagestream tag exclude from pruning

### PR DESCRIPTION
**Pull request by legionus**
Related RFE: https://bugzilla.redhat.com/show_bug.cgi?id=1410018

### CODE REVIEW

**Comment by miminar** [comment]
The regular expression can get very long and obfuscated.

Instead of adding new flag to the prune command. Could we allow to mark images and imagestream(tag)s as not prunable? For example with annotation like `openshift.io/image.noPrune=true`?

In case of the flag added to the prune command. I'd prefer the solution suggested by the customer - let the users specify a list of imagestream(tag)s in a file. I find the annotation approach more convenient though.

**Comment by bparees** [comment]
@miminar the annotation approach was discussed and rejected during planning and noted in the trello card:
"The better option would probably be to allow the imagestream itself to indicate it should be immune from pruning, but that's a big security problem (who's allowed to set that annotation/indicator?)"

we don't have good security models around annotation editing.  So this was the short path to provide the capability being requested.

The intent of the card was to include support for a file listing images to be excluded(in addition to the regex support), if that's not in this PR it should be added (I haven't reviewed it yet).

File: pkg/oc/admin/prune/images.go [hunk 1]
```diff
@@ -143,6 +145,7 @@ func NewCmdPruneImages(f *clientcmd.Factory, parentName, name string, out io.Wri
 	cmd.Flags().BoolVar(opts.AllImages, "all", *opts.AllImages, "Include images that were not pushed to the registry but have been mirrored by pullthrough.")
 	cmd.Flags().DurationVar(opts.KeepYoungerThan, "keep-younger-than", *opts.KeepYoungerThan, "Specify the minimum age of an image for it to be considered a candidate for pruning.")
 	cmd.Flags().IntVar(opts.KeepTagRevisions, "keep-tag-revisions", *opts.KeepTagRevisions, "Specify the number of image revisions for a tag in an image stream that will be preserved.")
+	cmd.Flags().StringVar(&opts.ExcludeImageStreamTag, "exclude-imagestream-tag", "", "The regular expression that will be applied to the ImageStream Tag to preserve it from pruning.")
```
Comment (bparees) [inline pkg/oc/admin/prune/images.go]: Add an arg that takes a file containing a list of regexes to exclude (probably won't be true regexes normally, will just be a list of image names, but no reason not to allow them to be regexes, unless you have performance concerns)

**Comment by legionus** [comment]
> The intent of the card was to include support for a file listing images to be excluded(in addition to the regex support), if that's not in this PR it should be added (I haven't reviewed it yet).

@bparees I thought we agreed to add an option with a regular expression and not a file.
Well. I'll add another option.

File: pkg/oc/admin/prune/images.go (hunk 1 above)
Comment (legionus) [reply pkg/oc/admin/prune/images.go]: @bparees Would it suit you if there are two options (this on and another with file as an argument) ?

**Comment by miminar** [comment]
> @miminar the annotation approach was discussed and rejected during planning and noted in the trello card:
> "The better option would probably be to allow the imagestream itself to indicate it should be immune from pruning, but that's a big security problem (who's allowed to set that annotation/indicator?)"

I see. I'm sorry for not participating in the discussion. I could imagine allowing at least a marking of images for those cluster-wide roles that already allow to edit images (e.g. `image-auditor`). I don't see security implications in that. Marking of imagestream(tag)s would be more challenging for sure.

File: pkg/oc/admin/prune/images.go (hunk 1 above)
Comment (bparees) [reply pkg/oc/admin/prune/images.go]: yes, two args is what i'd expect.

**Comment by bparees** [comment]
> @bparees I thought we agreed to add an option with a regular expression and not a file.

Sorry there was confusion, I have a different recollection but either way I think it speaks to the fact that we need to do a better job of updating the trello cards during planning or immediately after, with exactly what we intend to implement for the card (checklists are probably the best way to do this, rather than just putting it in the description).  I'll try to do better with that, but I definitely will need help from everyone else as well.

(those updates would also help QE drafting TCs and also bystanders who sometimes watch our cards and wonder what we're doing).

**Comment by bparees** [comment]
>  I could imagine allowing at least a marking of images for those cluster-wide roles that already allow to edit images (e.g. image-auditor). I don't see security implications in that. Marking of imagestream(tag)s would be more challenging for sure.

Sure, that seems reasonable.  However for now, this is exactly the behavior the customer requested in their RFE and I think it has some advantages (and some disadvantages, certainly) over having to mark the images, so i think it's worth doing and if we get additional requests for functionality in the future we can always do more.

File: pkg/oc/admin/prune/images.go (hunk 1 above)
Comment (legionus) [reply pkg/oc/admin/prune/images.go]: @bparees done

File: pkg/image/prune/prune.go [hunk 2]
```diff
@@ -260,21 +265,6 @@ func addImagesToGraph(g graph.Graph, images *imageapi.ImageList, algorithm prune
...
-				continue
-			}
-		}
-
-		age := metav1.Now().Sub(image.CreationTimestamp.Time)
-		if !algorithm.pruneOverSizeLimit && age < algorithm.keepYoungerThan {
-			glog.V(4).Infof("Image %q is younger than minimum pruning age, skipping (age=%v)", image.Name, age)
-			continue
-		}
-
```
Comment (miminar) [inline pkg/image/prune/prune.go]: So if these images were not added to the graph, the algorithm could remove blobs referenced by them, could it not?

File: pkg/oc/admin/prune/images.go [hunk 3]
```diff
@@ -143,6 +147,8 @@ func NewCmdPruneImages(f *clientcmd.Factory, parentName, name string, out io.Wri
...
+	cmd.Flags().StringVar(&opts.ExcludeImageStreamTagFile, "exclude-imagestream-tag-file", "", "The filename that contains the regular expressions that will be applied to the ImageStream Tag to preserve it from pruning.")
```
Comment (miminar) [inline pkg/oc/admin/prune/images.go]: `s/that will be applied to the ImageStream Tag to preserve it from pruning./matching ImageStreamTags excluded from pruning./`

Comment (miminar) [reply pkg/oc/admin/prune/images.go]: Add an example and point to regular expression syntax accepted.

File: pkg/oc/admin/prune/images.go (hunk 1 above)
Comment (miminar) [reply pkg/oc/admin/prune/images.go]: exclude-imagestreamtags?

File: pkg/oc/admin/prune/images.go (hunk 3 above)
Comment (miminar) [reply pkg/oc/admin/prune/images.go]: Also describe the string you're matching against (`<namespace>/<name>:<tag>`).

**Comment by miminar** [comment]
Also another extended test for the patterns file would be nice.

**Comment by legionus** [comment]
> Also another extended test for the patterns file would be nice.

@miminar Both of these options fill same array. If one works, then filtering works.

File: pkg/image/prune/prune.go (hunk 2 above)
Comment (legionus) [reply pkg/image/prune/prune.go]: @miminar yes. The layers will not have links in the graph.

**Comment by legionus** [comment]
@miminar comments addressed.

File: pkg/image/prune/prune.go (hunk 2 above)
Comment (miminar) [reply pkg/image/prune/prune.go]: That's a serious bug. And we need to back-port this pretty far. The mystery of blobs disappearing from the registry is solved!

Comment (miminar) [reply pkg/image/prune/prune.go]: Could you please capture the case of layers being removed when they should be kept in some unit/extended test? This is pretty important gap we should cover.

File: pkg/image/prune/prune.go [hunk 4]
```diff
@@ -335,17 +325,41 @@ func addImageStreamsToGraph(g graph.Graph, streams *imageapi.ImageStreamList, li
...
+						}
+					}
+				}
+
 				kind := oldImageRevisionReferenceKind
-				if algorithm.pruneOverSizeLimit {
-					if exceedsLimits(stream, imageNode.Image, limits) {
-						kind = WeakReferencedImageEdgeKind
+
+				if !istExcluded {
```
Comment (dmage) [inline pkg/image/prune/prune.go]:
```
if istExcluded {
   ...
} else if algorithm.pruneOverSizeLimit {
   ...
} else {
```

Do we want to add log messages for other cases?

File: pkg/image/prune/prune.go [hunk 5]
```diff
@@ -335,17 +325,41 @@ func addImageStreamsToGraph(g graph.Graph, streams *imageapi.ImageStreamList, li
...
+						if i < algorithm.keepTagRevisions {
+							kind = ReferencedImageEdgeKind
+						}
 					}
 				} else {
-					if i < algorithm.keepTagRevisions {
-						kind = ReferencedImageEdgeKind
-					}
+					glog.V(4).Infof("ImageStreamTag %q is excluded by the regular expression", dockerImageReference)
+					kind = ReferencedImageEdgeKind
```
Comment (dmage) [inline pkg/image/prune/prune.go]: Shouldn't it be WeakReferencedImageEdgeKind?

File: pkg/oc/admin/prune/images.go [hunk 6]
```diff
@@ -74,6 +76,10 @@ var (
 	  # To actually perform the prune operation, the confirm flag must be appended
 	  %[1]s %[2]s --keep-tag-revisions=3 --keep-younger-than=60m --confirm
 
+	  # To exclude an imagestream from pruning you can specify a regular expression
+	  # which will be applied to the '<namespace>/<name>:<tag>' string.
+	  %[1]s %[2]s --exclude-imagestreamtag=myproject/hello-openshift
```
Comment (dmage) [inline pkg/oc/admin/prune/images.go]: It may be worth mentioning the anchoring (^ and $).

File: pkg/oc/admin/prune/images.go [hunk 7]
```diff
@@ -625,3 +648,35 @@ func getRegistryClient(clientConfig *restclient.Config, registryCABundle string,
...
+func parsePatternsFile(filename string) ([]*regexp.Regexp, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var patterns []*regexp.Regexp
+
+	scanner := bufio.NewScanner(file)
```
Comment (dmage) [inline pkg/oc/admin/prune/images.go]: Don't use bufio.NewScanner for reading lines. It limits the length of tokens (iow. lines).
Use `file.ReadString('\n')` instead.

File: pkg/oc/admin/prune/images.go (hunk 6 above)
Comment (dmage) [reply pkg/oc/admin/prune/images.go]: Or at least change the example to `--exclude-imagestreamtag=^myproject/hello-openshift:`

Comment (dmage) [reply pkg/oc/admin/prune/images.go]: And I'd change the option name `--keep-imagestreamtag=^myproject/hello-openshift:`

File: pkg/image/prune/prune.go (hunk 5 above)
Comment (legionus) [reply pkg/image/prune/prune.go]: @dmage No. https://github.com/legionus/origin/blob/304d2b846cbaecc6b30a5b692d9b343cf2c95660/pkg/image/prune/prune.go#L635

Comment (dmage) [reply pkg/image/prune/prune.go]: OK, I thought excluded=pruned. s/excluded/protected/ or kept.

**Comment by bparees** [comment]
lgtm as well but will need to hold for 3.8 at this point.

**Comment by miminar** [comment]
Resolves #14863

**Comment by miminar** [comment]
@bparees we need the fix in 3.7 and most of earlier versions. Would you like to extract the fix from the exclusion flags in order to merge early?

File: pkg/image/prune/prune.go (hunk 2 above)
Comment (dmage) [reply pkg/image/prune/prune.go]: @miminar I'm on it. BZ 1487408

**Comment by bparees** [comment]
> @bparees we need the fix in 3.7 and most of earlier versions. Would you like to extract the fix from the exclusion flags in order to merge early?

@miminar  @legionus yeah, as I noted in the other bug (https://bugzilla.redhat.com/show_bug.cgi?id=1487408), that is what we need to do here (separate out the fix+backport it)

**Comment by bparees** [comment]
@legionus please give this a rebase and we can merge it.

**Comment by bparees** [comment]
@legionus bump, needs rebase.

**Comment by legionus** [comment]
@bparees on my way

**Comment by legionus** [comment]
Test `extended_image_registry` is broken:
```
Failure summary:


  1. Hosts:    localhost
     Play:     Create OpenShift certificates for master hosts
     Task:     Set Default scheduler predicates and priorities
     Message:  An unhandled exception occurred while running the lookup plugin 'openshift_master_facts_default_predicates'. Error was a <class 'ansible.errors.AnsibleError'>, original message: Unknown short_version 3.8
```

**Comment by bparees** [comment]
@legionus i worked with @stevekuznetsov on that this morning, it should be resolved now, hopefully the current run will pass.

**Comment by mfojtik** [comment]
@bparees i'm going to revert this commit because registry/util was removed from rebase branch so you can't use that helper anymore. Second option would be to disable the test.

**Comment by mfojtik** [comment]
See: https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/17115/test_pull_request_origin_extended_networking_minimal/10598/

**Comment by mfojtik** [comment]
@bparees i reverted this but realized with Alex that we moved that registry helper func into different package, so you just need to remove that import and strip the `registryutil.` in the test. Anyway, can you please resubmit this after the rebase lands? I really hate to rebase 80 commits to adapt this to post-rebase world.

**Comment by deads2k** [comment]
The use a regex here appears to make the command hard to use.  In addition, if I'm reading the code correctly, the caller either has to specify specific namespaces in his regex to avoid accidentally matching image streams owned by other people that should be pruned.  

Given those two things, why not just make it official and simply making some namespaces immune to pruning for a command.  That could be done via label selection of namespaces (users can't label namespaces), it wouldn't require an additional flag, and it wouldn't be vulnerable to less than obvious matching problems.

**Comment by bparees** [comment]
> The use a regex here appears to make the command hard to use.

if you're doing this, you have a need for the complexity.

>  In addition, if I'm reading the code correctly, the caller either has to specify specific namespaces in his regex to avoid accidentally matching image streams owned by other people that should be pruned.

yup.

> Given those two things, why not just make it official and simply making some namespaces immune to pruning for a command. 

"some namespaces"?  what namespaces?

>  That could be done via label selection of namespaces (users can't label namespaces), it wouldn't require an additional flag, and it wouldn't be vulnerable to less than obvious matching problems.

because you're assuming that the exclusion is scoped to a particular namespace.  What if i want to prune some but not all images in a particular namespace?  

This feature was implemented based on the specific request of a customer in fulfillment of how they wanted it to work.  You can read the RFE.

**Comment by bparees** [comment]
fyi i'm deferring this suggestion.  @legionus implemented sufficient functionality to address the customer need and we're moving on, i'm not willing to waste his time rewriting it when we have more pressing concerns.

Certainly in the future we can introduce the ability to exclude pruning by a namespace label.  Providing this (more flexible) capability today does not preclude providing that capability in the future.

once the rebase has landed we'll be re-merging this feature.

**Comment by smarterclayton** [comment]
I'm ok with this as it is - I understand the improvement David wants to make, I think it's a good long term thing to do, but I don't see a justification to rewrite it.

**Comment by miminar** [comment]
> Also, this is good design feedback that would be better earlier in the process. What can we do to make these suggestions earlier?

Well, the suggestion [was raised](https://github.com/openshift/origin/pull/16580#issuecomment-332539286) much earlier and was discussed in planning as well but was turned down.

**Comment by deads2k** [comment]
> I'm ok with this as it is - I understand the improvement David wants to make, I think it's a good long term thing to do, but I don't see a justification to go with that approach over the current approach that is worth rewriting over.

Just so we can avoid the concern about cost, I put together a pull which uses label and field selection to protect image.  It is complete, including tests, without any API changes here: https://github.com/openshift/origin/pull/17511.


> Also, this is good design feedback that would be better earlier in the process. What can we do to make these suggestions earlier?

I'm sure it was an oversight, but I don't see @juanvallejo, @fabianofranz, @openshift/cli-review, or @openshift/api-review tagged in this pull or in the bugzilla or in the trello card for review or comment.  I suspect that any of those actors are familiar with the over-arching design principles of `kubectl` and `oc` regarding object selection via the REST API and the CLI tools.

**Comment by ctrought** [comment]
/reopen

PR was reverted, I don't think any more granularity has been provided from any other options yet. Would still like to see more control over pruning image tags.

**Pull request merged**